lib/github_changelog_generator/fetcher.rb

```diff
@@ -0,0 +1,191 @@
+module GitHubChangelogGenerator
+  # A Fetcher responsible for all requests to GitHub and all basic manipulation with related data (such as filtering, validating, e.t.c)
+  #
+  # Example:
+  # fetcher = GitHubChangelogGenerator::Fetcher.new options
+  class Fetcher
+    PER_PAGE_NUMBER = 30
+    GH_RATE_LIMIT_EXCEEDED_MSG = "Warning: GitHub API rate limit (5000 per hour) exceeded, change log may be " \
+        "missing some issues. You can limit the number of issues fetched using the `--max-issues NUM` argument."
+
+    def initialize(options = {})
+      @options = options
+
+      @user = @options[:user]
+      @project = @options[:project]
+      @github_token = fetch_github_token
+      @tag_times_hash = {}
+      github_options = { per_page: PER_PAGE_NUMBER }
+      github_options[:oauth_token] = @github_token unless @github_token.nil?
+      github_options[:endpoint] = options[:github_endpoint] unless options[:github_endpoint].nil?
+      github_options[:site] = options[:github_endpoint] unless options[:github_site].nil?
+
+      begin
+        @github = Github.new github_options
+      rescue
+        puts GH_RATE_LIMIT_EXCEEDED_MSG.yellow
+      end
+    end
+
+    # Returns GitHub token. First try to use variable, provided by --token option, otherwise try to fetch it from CHANGELOG_GITHUB_TOKEN env variable.
+    #
+    # @return [String]
+    def fetch_github_token
+      env_var = @options[:token] ? @options[:token] : (ENV.fetch "CHANGELOG_GITHUB_TOKEN", nil)
+
+      unless env_var
+        puts "Warning: No token provided (-t option) and variable $CHANGELOG_GITHUB_TOKEN was not found.".yellow
+        puts "This script can make only 50 requests to GitHub API per hour without token!".yellow
+      end
+
+      env_var
+    end
+
+    # Fetch all tags from repo
+    # @return [Array] array of tags
+    def get_all_tags
+      if @options[:verbose]
+        print "Fetching tags...\r"
+      end
+
+      tags = []
+
+      begin
+        response = @github.repos.tags @options[:user], @options[:project]
+        page_i = 0
+        count_pages = response.count_pages
+        response.each_page do |page|
+          page_i += PER_PAGE_NUMBER
+          print "Fetching tags... #{page_i}/#{count_pages * PER_PAGE_NUMBER}\r"
+          tags.concat(page)
+        end
+        print "                               \r"
+
+        if tags.count == 0
+          puts "Warning: Can't find any tags in repo. Make sure, that you push tags to remote repo via 'git push --tags'".yellow
+        elsif @options[:verbose]
+          puts "Found #{tags.count} tags"
+        end
+
+      rescue
+        puts GH_RATE_LIMIT_EXCEEDED_MSG.yellow
+      end
+
+      tags
+    end
+
+    # This method fetch all closed issues and separate them to pull requests and pure issues
+    # (pull request is kind of issue in term of GitHub)
+    # @return [Tuple] with issues and pull requests
+    def fetch_issues_and_pull_requests
```

Comment:
Assignment Branch Condition size for fetch_issues_and_pull_requests is too high. [25.24/15]